codegen/smithy-go-codegen/src/test/java/software/amazon/smithy/go/codegen/GoJmespathExpressionGeneratorTest.java

```diff
@@ -138,24 +153,343 @@ public void testKeysFunctionExpression() {
-                v1 := v
-                v2 := v1.Key
-                v3 = append(v3, v2)
+                    v4 = append(v4, v...)
+                }
+                var v5 []string
+                for _, v := range v4 {
+                    v6 := v.InnerKey
+                    if v6 != nil {
+                        v5 = append(v5, *v6)
+                    }
                 }
                 """));
     }
+
+    @Test
+    public void testLengthFunctionExpression() {
+        var expr = "length(objectList)";
+
+        var writer = testWriter();
+        var generator = new GoJmespathExpressionGenerator(testContext(), writer);
+        var actual = generator.generate(JmespathExpression.parse(expr), new GoJmespathExpressionGenerator.Variable(
+                TEST_MODEL.expectShape(ShapeId.from("smithy.go.test#Struct")),
+                "input"
+        ));
+        assertThat(actual.shape(), Matchers.equalTo(ShapeUtil.INT_SHAPE));
+        assertThat(actual.ident(), Matchers.equalTo("v2"));
+        assertThat(writer.toString(), Matchers.containsString("""
+                v1 := input.ObjectList
+                v2 := len(v1)
+                """));
+    }
+
+    @Test
+    public void testLengthFunctionStringPtr() {
+        var expr = "length(simpleShape)";
+
+        var writer = testWriter();
+        var generator = new GoJmespathExpressionGenerator(testContext(), writer);
+        var actual = generator.generate(JmespathExpression.parse(expr), new GoJmespathExpressionGenerator.Variable(
+                TEST_MODEL.expectShape(ShapeId.from("smithy.go.test#Struct")),
+                "input"
+        ));
+        assertThat(actual.shape(), Matchers.equalTo(ShapeUtil.INT_SHAPE));
+        assertThat(actual.ident(), Matchers.equalTo("v2"));
+        assertThat(writer.toString(), Matchers.containsString("""
+                v1 := input.SimpleShape
+                var _v1 string
+                if v1 != nil {
+                    _v1 = *v1
+                }
+                v2 := len(_v1)
+                """));
+    }
+
+    @Test
+    public void testComparatorInt() {
+        var expr = "length(objectList) > `0`";
+
+        var writer = testWriter();
+        var generator = new GoJmespathExpressionGenerator(testContext(), writer);
+        var actual = generator.generate(JmespathExpression.parse(expr), new GoJmespathExpressionGenerator.Variable(
+                TEST_MODEL.expectShape(ShapeId.from("smithy.go.test#Struct")),
+                "input"
+        ));
+        assertThat(actual.shape(), Matchers.equalTo(ShapeUtil.BOOL_SHAPE));
+        assertThat(actual.ident(), Matchers.equalTo("v4"));
+        assertThat(writer.toString(), Matchers.containsString("""
+                v1 := input.ObjectList
+                v2 := len(v1)
+                v3 := 0
+                v4 := int64(v2) > int64(v3)
+                """));
+    }
+
+    @Test
+    public void testComparatorStringLHSNil() {
```

Comment:
Shapes are nilable by default for cross-language compatibility - so since `output.nested.nestedField` is just a plain `String` shape, we generate it as nilable.